[Bug] Got decompress error in ZstdStreamDecompressor

### Component Selection

- [x] Core Engine (Expression eval, Memory, Vector)
- [ ] Connectors / File Formats (Hive, Parquet, etc.)
- [ ] API / Bindings (Python, etc.)
- [ ] Build
- [ ] Other

### Describe the Bug

After #180 merged, some workload got Failed in ZSTD_DECOMPRESSOR.

### Reproduction Steps

Reproduced when:
1. using RowBased Shuffle.
2. compressed size is large than origin raw data size.

### Bolt Version / Commit ID

d748fb5

### System Configuration

```markdown
- **OS**: (e.g. Ubuntu 22.04, CentOS 7)
- **Compiler**: (e.g. GCC 11, Clang 14)
- **Build Type**: (Debug / Release / RelWithDebInfo)
- **CPU Arch**: (e.g. x86_64 AVX2, ARM64)
- **Framework**: (e.g. Spark 3.3, PrestoDB)
```

### Logs / Stack Trace

```shell
Job aborted due to stage failure: Task 58 in stage 14.0 failed 8 times, most recent failure: Lost task 58.7 in stage 14.0 (TID 796) (n175-090-199.byted.org executor 28): org.apache.gluten.exception.GlutenException: org.apache.gluten.exception.GlutenException: Exception: BoltRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Operator::getOutput failed for [operator: ValueStream, plan node ID: 0]: Error during calling Java code from native code: org.apache.gluten.exception.GlutenException: org.apache.gluten.exception.GlutenException: Exception: BoltRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Failed in ZSTD_DECOMPRESSOR: 
Retriable: False
Expression: false
Context: Operator: SparkShuffleReader[0] 0
Additional Context: Operator: ValueStream[0] 0
Function: decompress
File: /home/xiaofeng.zhang/Workspace/bolt-internal/bolt/shuffle/sparksql/compression/ZstdStreamCodec.cpp
Line: 169
Stack trace:
# 0  _ZN9bytedance4bolt7process10StackTraceC1Ei
# 1  _ZN9bytedance4bolt13BoltExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN9bytedance4bolt6detail13boltCheckFailINS0_16BoltRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_17BoltCheckFailArgsET0_
# 3  _ZN9bytedance4bolt7shuffle8sparksql22ZstdStreamDecompressor10decompressEPKhlPhl
# 4  _ZN9bytedance4bolt7shuffle8sparksql25AdaptiveParallelZstdCodec10DecompressEPN5arrow2io11InputStreamEPhiiRiRbSA_RNS2_15RowVectorLayoutE
# 5  _ZN9bytedance4bolt7shuffle8sparksql15RowBlockPayload11deserializeEPN5arrow2io11InputStreamEPhiRiPNS2_25AdaptiveParallelZstdCodecERSt6vectorISt17basic_string_viewIcSt11char_traitsIcEESaISG_EES9_RbSK_RNS2_15RowVectorLayoutERm
# 6  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer12nextFromRowsEv
# 7  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer4nextEv
# 8  _ZN9bytedance4bolt7shuffle8sparksql18SparkShuffleReader9getOutputEv
# 9  _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 10 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 11 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 12 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 13 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 14 _ZN6gluten24WholeStageResultIterator4nextEv
# 15 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 16 0x00007f350c3e20db
# 17 0x00007f3504ff1904
# 18 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 19 _ZL20jni_invoke_nonstaticP7JNIEnv_P9JavaValueP8_jobject11JNICallTypeP10_jmethodIDP18JNI_ArgumentPusherP10JavaThread.constprop.1
# 20 jni_CallBooleanMethodV
# 21 _ZN7JNIEnv_17CallBooleanMethodEP8_jobjectP10_jmethodIDz
# 22 _ZNK6gluten24JniColumnarBatchIterator12nextInternalEv
# 23 _ZN6gluten24JniColumnarBatchIterator4nextEv
# 24 _ZN6gluten11ValueStream9getOutputEv
# 25 _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 26 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 27 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 28 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 29 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 30 _ZN6gluten24WholeStageResultIterator4nextEv
# 31 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 32 0x00007f350c3e20db
# 33 0x00007f3504ff1904
# 34 0x00007f350b8c423e
# 35 0x00007f350b8c423e
# 36 0x00007f350b8c423e
# 37 0x00007f350b8c4397
# 38 0x00007f350b8c4397
# 39 0x00007f350b8c4397
# 40 0x00007f350b8c4397
# 41 0x00007f350b8c4397
# 42 0x00007f350b8c47fb
# 43 0x00007f350b8c4397
# 44 0x00007f350b8c46a2
# 45 0x00007f350b8c423e
# 46 0x00007f350b8c46a2
# 47 0x00007f350b8bbcc9
# 48 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 49 _ZN9JavaCalls12call_virtualEP9JavaValue6HandleP5KlassP6SymbolS6_P10JavaThread
# 50 _ZL12thread_entryP10JavaThreadS0_
# 51 _ZN10JavaThread17thread_main_innerEv
# 52 _ZN6Thread8call_runEv
# 53 _ZL19thread_native_entryP6Thread
# 54 start_thread
# 55 clone

	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:42)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
	at org.apache.gluten.iterator.IteratorsV1$InvocationFlowProtection.hasNext(IteratorsV1.scala:154)
	at org.apache.gluten.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:66)
	at org.apache.gluten.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:38)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.hasNext(IteratorsV1.scala:95)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:32)
	at org.apache.gluten.vectorized.ColumnarBatchInIterator.hasNext(ColumnarBatchInIterator.java:36)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.hasNext0(ColumnarBatchOutIterator.java:58)
	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:37)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
	at org.apache.gluten.iterator.IteratorsV1$InvocationFlowProtection.hasNext(IteratorsV1.scala:154)
	at org.apache.gluten.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:66)
	at org.apache.gluten.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:38)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.hasNext(IteratorsV1.scala:95)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.combinedWrite(ColumnarShuffleWriter.scala:215)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.internalWrite(ColumnarShuffleWriter.scala:267)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.write(ColumnarShuffleWriter.scala:387)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:61)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:134)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:544)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1642)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:547)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.apache.gluten.exception.GlutenException: Exception: BoltRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Failed in ZSTD_DECOMPRESSOR: 
Retriable: False
Expression: false
Context: Operator: SparkShuffleReader[0] 0
Additional Context: Operator: ValueStream[0] 0
Function: decompress
File: /home/xiaofeng.zhang/Workspace/bolt-internal/bolt/shuffle/sparksql/compression/ZstdStreamCodec.cpp
Line: 169
Stack trace:
# 0  _ZN9bytedance4bolt7process10StackTraceC1Ei
# 1  _ZN9bytedance4bolt13BoltExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN9bytedance4bolt6detail13boltCheckFailINS0_16BoltRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_17BoltCheckFailArgsET0_
# 3  _ZN9bytedance4bolt7shuffle8sparksql22ZstdStreamDecompressor10decompressEPKhlPhl
# 4  _ZN9bytedance4bolt7shuffle8sparksql25AdaptiveParallelZstdCodec10DecompressEPN5arrow2io11InputStreamEPhiiRiRbSA_RNS2_15RowVectorLayoutE
# 5  _ZN9bytedance4bolt7shuffle8sparksql15RowBlockPayload11deserializeEPN5arrow2io11InputStreamEPhiRiPNS2_25AdaptiveParallelZstdCodecERSt6vectorISt17basic_string_viewIcSt11char_traitsIcEESaISG_EES9_RbSK_RNS2_15RowVectorLayoutERm
# 6  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer12nextFromRowsEv
# 7  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer4nextEv
# 8  _ZN9bytedance4bolt7shuffle8sparksql18SparkShuffleReader9getOutputEv
# 9  _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 10 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 11 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 12 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 13 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 14 _ZN6gluten24WholeStageResultIterator4nextEv
# 15 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 16 0x00007f350c3e20db
# 17 0x00007f3504ff1904
# 18 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 19 _ZL20jni_invoke_nonstaticP7JNIEnv_P9JavaValueP8_jobject11JNICallTypeP10_jmethodIDP18JNI_ArgumentPusherP10JavaThread.constprop.1
# 20 jni_CallBooleanMethodV
# 21 _ZN7JNIEnv_17CallBooleanMethodEP8_jobjectP10_jmethodIDz
# 22 _ZNK6gluten24JniColumnarBatchIterator12nextInternalEv
# 23 _ZN6gluten24JniColumnarBatchIterator4nextEv
# 24 _ZN6gluten11ValueStream9getOutputEv
# 25 _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 26 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 27 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 28 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 29 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 30 _ZN6gluten24WholeStageResultIterator4nextEv
# 31 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 32 0x00007f350c3e20db
# 33 0x00007f3504ff1904
# 34 0x00007f350b8c423e
# 35 0x00007f350b8c423e
# 36 0x00007f350b8c423e
# 37 0x00007f350b8c4397
# 38 0x00007f350b8c4397
# 39 0x00007f350b8c4397
# 40 0x00007f350b8c4397
# 41 0x00007f350b8c4397
# 42 0x00007f350b8c47fb
# 43 0x00007f350b8c4397
# 44 0x00007f350b8c46a2
# 45 0x00007f350b8c423e
# 46 0x00007f350b8c46a2
# 47 0x00007f350b8bbcc9
# 48 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 49 _ZN9JavaCalls12call_virtualEP9JavaValue6HandleP5KlassP6SymbolS6_P10JavaThread
# 50 _ZL12thread_entryP10JavaThreadS0_
# 51 _ZN10JavaThread17thread_main_innerEv
# 52 _ZN6Thread8call_runEv
# 53 _ZL19thread_native_entryP6Thread
# 54 start_thread
# 55 clone

	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.hasNext0(ColumnarBatchOutIterator.java:58)
	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:37)
	... 32 more

Retriable: False
Function: operator()
File: /home/xiaofeng.zhang/Workspace/bolt-internal/bolt/exec/Driver.cpp
Line: 688
Stack trace:
# 0  _ZN9bytedance4bolt7process10StackTraceC1Ei
# 1  _ZN9bytedance4bolt13BoltExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN9bytedance4bolt6detail13boltCheckFailINS0_16BoltRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_17BoltCheckFailArgsET0_
# 3  _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv.cold
# 4  _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 5  _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 6  _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 7  _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 8  _ZN6gluten24WholeStageResultIterator4nextEv
# 9  Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 10 0x00007f350c3e20db
# 11 0x00007f3504ff1904
# 12 0x00007f350b8c423e
# 13 0x00007f350b8c423e
# 14 0x00007f350b8c423e
# 15 0x00007f350b8c4397
# 16 0x00007f350b8c4397
# 17 0x00007f350b8c4397
# 18 0x00007f350b8c4397
# 19 0x00007f350b8c4397
# 20 0x00007f350b8c47fb
# 21 0x00007f350b8c4397
# 22 0x00007f350b8c46a2
# 23 0x00007f350b8c423e
# 24 0x00007f350b8c46a2
# 25 0x00007f350b8bbcc9
# 26 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 27 _ZN9JavaCalls12call_virtualEP9JavaValue6HandleP5KlassP6SymbolS6_P10JavaThread
# 28 _ZL12thread_entryP10JavaThreadS0_
# 29 _ZN10JavaThread17thread_main_innerEv
# 30 _ZN6Thread8call_runEv
# 31 _ZL19thread_native_entryP6Thread
# 32 start_thread
# 33 clone

	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:42)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
	at org.apache.gluten.iterator.IteratorsV1$InvocationFlowProtection.hasNext(IteratorsV1.scala:154)
	at org.apache.gluten.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:66)
	at org.apache.gluten.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:38)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.hasNext(IteratorsV1.scala:95)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.combinedWrite(ColumnarShuffleWriter.scala:215)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.internalWrite(ColumnarShuffleWriter.scala:267)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.write(ColumnarShuffleWriter.scala:387)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:61)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:134)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:544)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1642)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:547)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.apache.gluten.exception.GlutenException: Exception: BoltRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Operator::getOutput failed for [operator: ValueStream, plan node ID: 0]: Error during calling Java code from native code: org.apache.gluten.exception.GlutenException: org.apache.gluten.exception.GlutenException: Exception: BoltRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Failed in ZSTD_DECOMPRESSOR: 
Retriable: False
Expression: false
Context: Operator: SparkShuffleReader[0] 0
Additional Context: Operator: ValueStream[0] 0
Function: decompress
File: /home/xiaofeng.zhang/Workspace/bolt-internal/bolt/shuffle/sparksql/compression/ZstdStreamCodec.cpp
Line: 169
Stack trace:
# 0  _ZN9bytedance4bolt7process10StackTraceC1Ei
# 1  _ZN9bytedance4bolt13BoltExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN9bytedance4bolt6detail13boltCheckFailINS0_16BoltRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_17BoltCheckFailArgsET0_
# 3  _ZN9bytedance4bolt7shuffle8sparksql22ZstdStreamDecompressor10decompressEPKhlPhl
# 4  _ZN9bytedance4bolt7shuffle8sparksql25AdaptiveParallelZstdCodec10DecompressEPN5arrow2io11InputStreamEPhiiRiRbSA_RNS2_15RowVectorLayoutE
# 5  _ZN9bytedance4bolt7shuffle8sparksql15RowBlockPayload11deserializeEPN5arrow2io11InputStreamEPhiRiPNS2_25AdaptiveParallelZstdCodecERSt6vectorISt17basic_string_viewIcSt11char_traitsIcEESaISG_EES9_RbSK_RNS2_15RowVectorLayoutERm
# 6  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer12nextFromRowsEv
# 7  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer4nextEv
# 8  _ZN9bytedance4bolt7shuffle8sparksql18SparkShuffleReader9getOutputEv
# 9  _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 10 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 11 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 12 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 13 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 14 _ZN6gluten24WholeStageResultIterator4nextEv
# 15 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 16 0x00007f350c3e20db
# 17 0x00007f3504ff1904
# 18 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 19 _ZL20jni_invoke_nonstaticP7JNIEnv_P9JavaValueP8_jobject11JNICallTypeP10_jmethodIDP18JNI_ArgumentPusherP10JavaThread.constprop.1
# 20 jni_CallBooleanMethodV
# 21 _ZN7JNIEnv_17CallBooleanMethodEP8_jobjectP10_jmethodIDz
# 22 _ZNK6gluten24JniColumnarBatchIterator12nextInternalEv
# 23 _ZN6gluten24JniColumnarBatchIterator4nextEv
# 24 _ZN6gluten11ValueStream9getOutputEv
# 25 _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 26 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 27 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 28 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 29 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 30 _ZN6gluten24WholeStageResultIterator4nextEv
# 31 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 32 0x00007f350c3e20db
# 33 0x00007f3504ff1904
# 34 0x00007f350b8c423e
# 35 0x00007f350b8c423e
# 36 0x00007f350b8c423e
# 37 0x00007f350b8c4397
# 38 0x00007f350b8c4397
# 39 0x00007f350b8c4397
# 40 0x00007f350b8c4397
# 41 0x00007f350b8c4397
# 42 0x00007f350b8c47fb
# 43 0x00007f350b8c4397
# 44 0x00007f350b8c46a2
# 45 0x00007f350b8c423e
# 46 0x00007f350b8c46a2
# 47 0x00007f350b8bbcc9
# 48 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 49 _ZN9JavaCalls12call_virtualEP9JavaValue6HandleP5KlassP6SymbolS6_P10JavaThread
# 50 _ZL12thread_entryP10JavaThreadS0_
# 51 _ZN10JavaThread17thread_main_innerEv
# 52 _ZN6Thread8call_runEv
# 53 _ZL19thread_native_entryP6Thread
# 54 start_thread
# 55 clone

	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:42)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
	at org.apache.gluten.iterator.IteratorsV1$InvocationFlowProtection.hasNext(IteratorsV1.scala:154)
	at org.apache.gluten.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:66)
	at org.apache.gluten.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:38)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.hasNext(IteratorsV1.scala:95)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:32)
	at org.apache.gluten.vectorized.ColumnarBatchInIterator.hasNext(ColumnarBatchInIterator.java:36)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.hasNext0(ColumnarBatchOutIterator.java:58)
	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:37)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
	at org.apache.gluten.iterator.IteratorsV1$InvocationFlowProtection.hasNext(IteratorsV1.scala:154)
	at org.apache.gluten.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:66)
	at org.apache.gluten.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:38)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.hasNext(IteratorsV1.scala:95)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.gluten.backendsapi.bolt.WholeStageIteratorWrapper.hasNext(BoltIteratorApi.scala:71)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.combinedWrite(ColumnarShuffleWriter.scala:215)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.internalWrite(ColumnarShuffleWriter.scala:267)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.write(ColumnarShuffleWriter.scala:387)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:61)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:134)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:544)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1642)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:547)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.apache.gluten.exception.GlutenException: Exception: BoltRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Failed in ZSTD_DECOMPRESSOR: 
Retriable: False
Expression: false
Context: Operator: SparkShuffleReader[0] 0
Additional Context: Operator: ValueStream[0] 0
Function: decompress
File: /home/xiaofeng.zhang/Workspace/bolt-internal/bolt/shuffle/sparksql/compression/ZstdStreamCodec.cpp
Line: 169
Stack trace:
# 0  _ZN9bytedance4bolt7process10StackTraceC1Ei
# 1  _ZN9bytedance4bolt13BoltExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN9bytedance4bolt6detail13boltCheckFailINS0_16BoltRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_17BoltCheckFailArgsET0_
# 3  _ZN9bytedance4bolt7shuffle8sparksql22ZstdStreamDecompressor10decompressEPKhlPhl
# 4  _ZN9bytedance4bolt7shuffle8sparksql25AdaptiveParallelZstdCodec10DecompressEPN5arrow2io11InputStreamEPhiiRiRbSA_RNS2_15RowVectorLayoutE
# 5  _ZN9bytedance4bolt7shuffle8sparksql15RowBlockPayload11deserializeEPN5arrow2io11InputStreamEPhiRiPNS2_25AdaptiveParallelZstdCodecERSt6vectorISt17basic_string_viewIcSt11char_traitsIcEESaISG_EES9_RbSK_RNS2_15RowVectorLayoutERm
# 6  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer12nextFromRowsEv
# 7  _ZN9bytedance4bolt7shuffle8sparksql29BoltColumnarBatchDeserializer4nextEv
# 8  _ZN9bytedance4bolt7shuffle8sparksql18SparkShuffleReader9getOutputEv
# 9  _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 10 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 11 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 12 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 13 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 14 _ZN6gluten24WholeStageResultIterator4nextEv
# 15 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 16 0x00007f350c3e20db
# 17 0x00007f3504ff1904
# 18 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 19 _ZL20jni_invoke_nonstaticP7JNIEnv_P9JavaValueP8_jobject11JNICallTypeP10_jmethodIDP18JNI_ArgumentPusherP10JavaThread.constprop.1
# 20 jni_CallBooleanMethodV
# 21 _ZN7JNIEnv_17CallBooleanMethodEP8_jobjectP10_jmethodIDz
# 22 _ZNK6gluten24JniColumnarBatchIterator12nextInternalEv
# 23 _ZN6gluten24JniColumnarBatchIterator4nextEv
# 24 _ZN6gluten11ValueStream9getOutputEv
# 25 _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv
# 26 _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 27 _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 28 _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 29 _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 30 _ZN6gluten24WholeStageResultIterator4nextEv
# 31 Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 32 0x00007f350c3e20db
# 33 0x00007f3504ff1904
# 34 0x00007f350b8c423e
# 35 0x00007f350b8c423e
# 36 0x00007f350b8c423e
# 37 0x00007f350b8c4397
# 38 0x00007f350b8c4397
# 39 0x00007f350b8c4397
# 40 0x00007f350b8c4397
# 41 0x00007f350b8c4397
# 42 0x00007f350b8c47fb
# 43 0x00007f350b8c4397
# 44 0x00007f350b8c46a2
# 45 0x00007f350b8c423e
# 46 0x00007f350b8c46a2
# 47 0x00007f350b8bbcc9
# 48 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 49 _ZN9JavaCalls12call_virtualEP9JavaValue6HandleP5KlassP6SymbolS6_P10JavaThread
# 50 _ZL12thread_entryP10JavaThreadS0_
# 51 _ZN10JavaThread17thread_main_innerEv
# 52 _ZN6Thread8call_runEv
# 53 _ZL19thread_native_entryP6Thread
# 54 start_thread
# 55 clone

	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.hasNext0(ColumnarBatchOutIterator.java:58)
	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:37)
	... 32 more

Retriable: False
Function: operator()
File: /home/xiaofeng.zhang/Workspace/bolt-internal/bolt/exec/Driver.cpp
Line: 688
Stack trace:
# 0  _ZN9bytedance4bolt7process10StackTraceC1Ei
# 1  _ZN9bytedance4bolt13BoltExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN9bytedance4bolt6detail13boltCheckFailINS0_16BoltRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_17BoltCheckFailArgsET0_
# 3  _ZZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEEENKUlvE3_clEv.cold
# 4  _ZN9bytedance4bolt4exec6Driver11runInternalERSt10shared_ptrIS2_ERS3_INS1_13BlockingStateEERS3_INS0_9RowVectorEE
# 5  _ZN9bytedance4bolt4exec6Driver4nextERSt10shared_ptrINS1_13BlockingStateEE
# 6  _ZN9bytedance4bolt4exec4Task4nextEPN5folly10SemiFutureINS3_4UnitEEE
# 7  _ZN6gluten24WholeStageResultIterator12nextInternalEv
# 8  _ZN6gluten24WholeStageResultIterator4nextEv
# 9  Java_org_apache_gluten_vectorized_ColumnarBatchOutIterator_nativeHasNext
# 10 0x00007f350c3e20db
# 11 0x00007f3504ff1904
# 12 0x00007f350b8c423e
# 13 0x00007f350b8c423e
# 14 0x00007f350b8c423e
# 15 0x00007f350b8c4397
# 16 0x00007f350b8c4397
# 17 0x00007f350b8c4397
# 18 0x00007f350b8c4397
# 19 0x00007f350b8c4397
# 20 0x00007f350b8c47fb
# 21 0x00007f350b8c4397
# 22 0x00007f350b8c46a2
# 23 0x00007f350b8c423e
# 24 0x00007f350b8c46a2
# 25 0x00007f350b8bbcc9
# 26 _ZN9JavaCalls11call_helperEP9JavaValueRK12methodHandleP17JavaCallArgumentsP10JavaThread
# 27 _ZN9JavaCalls12call_virtualEP9JavaValue6HandleP5KlassP6SymbolS6_P10JavaThread
# 28 _ZL12thread_entryP10JavaThreadS0_
# 29 _ZN10JavaThread17thread_main_innerEv
# 30 _ZN6Thread8call_runEv
# 31 _ZL19thread_native_entryP6Thread
# 32 start_thread
# 33 clone

	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.hasNext0(ColumnarBatchOutIterator.java:58)
	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:37)
	... 21 more

Driver stacktrace:
```

### Expected Behavior

_No response_

### Additional context

_No response_

--- a/bolt/shuffle/sparksql/CompressionStream.h
+++ b/bolt/shuffle/sparksql/CompressionStream.h
@@ -317,6 +317,7 @@ class AdaptiveParallelZstdCodec {
           // stream
           bytedance::bolt::NanosecondTimer timer1(&writeTime);
           RETURN_NOT_OK(outputStream->Write(output, outputPos));
+          outputPos = 0;
         }
       } while (inputLen > 0);
     }

--- a/bolt/shuffle/sparksql/tests/AdaptiveParallelZstdCodecTest.cpp
+++ b/bolt/shuffle/sparksql/tests/AdaptiveParallelZstdCodecTest.cpp
@@ -115,6 +115,31 @@ void runRoundTrip(
   EXPECT_EQ(0, std::memcmp(output.data(), expected.data(), expected.size()));
 }
 
+std::vector<uint8_t> buildIncompressibleData(size_t size) {
+  std::vector<uint8_t> data(size);
+  uint32_t state = 0x12345678u;
+  for (size_t i = 0; i < size; ++i) {
+    state = state * 1664525u + 1013904223u;
+    data[i] = static_cast<uint8_t>(state >> 24);
+  }
+  return data;
+}
+
+std::vector<uint8_t> buildIncompressibleRow(size_t payloadSize, uint32_t seed) {
+  std::vector<uint8_t> row(sizeof(int32_t) + payloadSize);
+  auto payloadSize32 = static_cast<int32_t>(payloadSize);
+  std::memcpy(row.data(), &payloadSize32, sizeof(int32_t));
+
+  auto payload = buildIncompressibleData(payloadSize);
+  uint32_t state = seed;
+  for (size_t i = 0; i < payloadSize; ++i) {
+    state = state * 1103515245u + 12345u;
+    payload[i] ^= static_cast<uint8_t>(state >> 24);
+  }
+  std::memcpy(row.data() + sizeof(int32_t), payload.data(), payload.size());
+  return row;
+}
+
 } // namespace
 
 TEST(AdaptiveParallelZstdCodecTest, RoundTripSmallPayloads) {
@@ -142,6 +167,29 @@ TEST(AdaptiveParallelZstdCodecTest, RoundTripLargePayload) {
   }
 
   runRoundTrip(rows, rawSize, RowVectorLayout::kComposite);
+}
+
+TEST(
+    AdaptiveParallelZstdCodecTest,
+    CompressAndFlushStressRoundTripWithoutCorruption) {
+  constexpr int32_t kRounds = 6;
+  constexpr int32_t kRowsPerRound = 256;
+  const auto payloadSize =
+      static_cast<size_t>(ZSTD_CStreamInSize() - sizeof(int32_t) - 1);
+
+  for (int32_t round = 0; round < kRounds; ++round) {
+    std::vector<std::vector<uint8_t>> rows;
+    rows.reserve(kRowsPerRound);
+
+    int64_t rawSize = 0;
+    for (int32_t i = 0; i < kRowsPerRound; ++i) {
+      rows.emplace_back(buildIncompressibleRow(
+          payloadSize, static_cast<uint32_t>(round * kRowsPerRound + i + 1)));
+      rawSize += static_cast<int64_t>(rows.back().size());
+    }
+
+    runRoundTrip(rows, rawSize, RowVectorLayout::kComposite);
+  }
 }
 
 } // namespace bytedance::bolt::shuffle::sparksql::test

